rpc/http.go

```diff
@@ -0,0 +1,268 @@
+	"github.com/aurora-is-near/relayer2-base/utils"
+	"github.com/fasthttp/websocket"
+	"github.com/valyala/fasthttp"
+	"github.com/valyala/fasthttp/fasthttpadaptor"
+	"github.com/valyala/fasthttp/reuseport"
+)
+
+const (
+	maxRequestContentLength = 1024 * 1024 * 5
+	DefaultContentType      = "application/json"
+
+	wsReadBuffer       = 1024
+	wsWriteBuffer      = 1024
+	wsMessageSizeLimit = 15 * 1024 * 1024
+)
+
+// https://www.jsonrpc.org/historical/json-rpc-over-http.html#id13
+var acceptedContentTypes = []string{DefaultContentType, "application/json-rpc", "application/jsonrequest"}
+
+type HttpServer struct {
+	Logger     *log.Logger
+	Config     HttpConfig
+	resolver   Resolver
+	listener   net.Listener
+	wsUpgrader *websocket.FastHTTPUpgrader
+}
+
+// httpConfig holds both http and websocket configuration elements
+type HttpConfig struct {
+	HttpEndpoint       string
+	HttpCors           []string
+	HttpCompress       bool
+	HttpTimeout        time.Duration
+	WsEndpoint         string
+	WsHandshakeTimeout time.Duration
+	WsOnly             bool
+}
+
+var upgrader = websocket.FastHTTPUpgrader{
+	ReadBufferSize:  wsReadBuffer,
+	WriteBufferSize: wsWriteBuffer,
+	WriteBufferPool: new(sync.Pool),
+}
+
+// Run enables CORS and Compression handlers if enabled on configuration and
+// starts the fasthttp server
+func (h *HttpServer) Run(ctx context.Context, resolver Resolver) error {
+	h.Logger.Info().Msgf("starting HTTP server on %s", h.Config.HttpEndpoint)
+	var err error
+	h.listener, err = reuseport.Listen("tcp4", h.Config.HttpEndpoint)
+	if err != nil {
+		return err
+	}
+
+	h.resolver = resolver
+	h.wsUpgrader = &websocket.FastHTTPUpgrader{
+		HandshakeTimeout: h.Config.WsHandshakeTimeout * time.Second,
+		CheckOrigin: func(ctx *fasthttp.RequestCtx) bool {
+			return true
+		},
+	}
+
+	var reqHandler fasthttp.RequestHandler
+	// Use CorsHandler if CORS settings are applied
+	withCors := h.newCorsHandler(h.Config.HttpCors)
+	if withCors != nil {
+		reqHandler = withCors.CorsMiddleware(h.mainHandler)
+	} else {
+		reqHandler = h.mainHandler
+	}
+	// Transparently compress the response body if the request contains 'gzip' or 'deflate' 'Accept-Encoding' header
+	// and configuration is enabled
+	if h.Config.HttpCompress {
+		reqHandler = fasthttp.CompressHandler(reqHandler)
+	}
+	reqHandler = fasthttp.TimeoutHandler(reqHandler, h.Config.HttpTimeout*time.Second, "request timeout")
+
+	go func() {
+		if err := fasthttp.Serve(h.listener, reqHandler); err != nil {
+			h.Logger.Fatal().Msgf("error while serving http: %v", err)
+		}
+	}()
+	return nil
+}
+
+// mainHandler is the handler called by the fasthttp server when a proper http
+// request is received. It calls the appropriate handler based on transport type
+func (h *HttpServer) mainHandler(ctx *fasthttp.RequestCtx) {
+	r := &http.Request{}
+	if err := fasthttpadaptor.ConvertRequest(ctx, r, false); err != nil {
+		ctx.Error("Internal Server Error", fasthttp.StatusInternalServerError)
```

Comment:
As you mentioned no valuable text is returning but `ctx.Error` additionally resets the response which may be useful especially for fasthttp server. 

- we can either keep the current structure (since Error method sets the content type as `text/plain`)
- or update as
 ```
ctx.Response.Reset()
 ctx.SetStatusCode(statusCode) 
```

What do you think?

Reply:
what is the benefit of resetting response, at this point for example?

Reply:
sorry my bad, I previously misunderstood that resetting the response before assigning any new data was mandatory to prevent data corruption (because of using the buffers from resource pool). However, after reading the fasthttp documents again, I realized that it is mandatory if you only edit something before. 
will update according to your comment